binarytree/__init__.py

```diff
@@ -451,6 +466,126 @@ def _generate_values(height, multiplier=1):
+    :type: Int
+    """
+    root = _prepare_tree(root)
+    if not isinstance(index, int) or index < 0:
+        raise ValueError("Requested id must be a non-negative integer.")
+    current_nodes = [root]
+    current_id = -1
+    current_index = 0
+
+    while current_nodes and current_id < index:
+        next_nodes = []
+        current_index = 0
+
+        while current_index < len(current_nodes) and current_id < index:
+            node = current_nodes[current_index]
+            if node is not _null:
+                left_child = _left_of(node)
+                right_child = _right_of(node)
+            else:
+                left_child = _null
+                right_child = _null
+
+            if left_child != _null:
+                next_nodes.append(left_child)
+            else:
+                next_nodes.append(_null)
+            if right_child != _null:
+                next_nodes.append(right_child)
+            else:
+                next_nodes.append(_null)
+            current_index += 1
+            current_id += 1
+        current_nodes = next_nodes
+
+    if node is _null:
+        raise IndexError("Requested node id not present in tree.")
+    return node
+
+
+def _get_parent_of_node_with_id(root, index):
+    """Get the index in the tree of the parent of a node.
+    Used for setting new nodes with the setitem method.
+
+    :param root: The root node of the tree
+    :type: Node
+    :param index: index of the node whose parent you're looking for
+    :type: Int
+    """
+    root = _prepare_tree(root)
+    current_nodes = [root]
+    current_id = 0
+    current_index = 0
+
+    while current_nodes and current_id <= index:
+        next_nodes = []
+        current_index = 0
+
+        while current_index < len(current_nodes) and current_id <= index:
+            node = current_nodes[current_index]
+            if _is_node(node):
+                left_child = _left_of(node)
+                right_child = _right_of(node)
+            else:
+                left_child = _null
+                right_child = _null
+
+            if left_child != _null:
+                next_nodes.append(left_child)
+            elif _is_node(node):
+                next_nodes.append((current_id, 'left'))
+            else:
+                next_nodes.append(_null)
+            if right_child != _null:
+                next_nodes.append(right_child)
+            elif _is_node(node):
+                next_nodes.append((current_id, 'right'))
+            else:
+                next_nodes.append(_null)
+            current_index += 1
+            current_id += 1
+        current_nodes = next_nodes
+
+    if node is _null:
+        raise IndexError("Requested node id's parent not present in tree.")
+    elif _is_node(node):
+        return _parent_of(node), 'right' if node.is_right_child_of(_parent_of(node)) else 'left'
+    else:
+        return _get_node_with_id(root, node[0]), node[1]
+
+
+def _set_node_with_id(root, index, value):
```

Comment:
Consider these observations when using [list representation](https://en.wikipedia.org/wiki/Binary_tree#Arrays):

```
1) Given a node at index i, its left, and right children are stored at indices 2*i+1, 2*i+2 respectively.
```
``` 
2) All nodes with odd IDs are right nodes, and those with even IDs are left nodes.
```
Using these facts + the `_build_list_2` method, you can greatly simplify the method. You probably won't need `_get_parent_of_node_with_id` at all :)